src/Compilers/Core/Portable/SourceGeneration/Nodes/TransformNode.cs

```diff
@@ -44,23 +56,41 @@ public NodeStateTable<TOutput> UpdateStateTable(DriverStateTable.Builder builder
             // - Added: perform transform and add
             // - Modified: perform transform and do element wise comparison with previous results
 
-            var newTable = previousTable.ToBuilder();
+            var newTable = previousTable.ToBuilder(builder.DriverState.TrackIncrementalSteps);
 
             foreach (var entry in sourceTable)
             {
-                if (entry.state == EntryState.Removed)
+                if (entry.State == EntryState.Removed && newTable.TryRemoveEntries())
+                {
+                    if (newTable.TrackIncrementalSteps)
+                    {
+                        newTable.RecordStepInfoForLastEntry(_name, TimeSpan.Zero, ImmutableArray.Create((entry.Step!, entry.OutputIndex)), EntryState.Removed);
+                    }
+                }
+                else if (entry.State == EntryState.Cached && newTable.TryUseCachedEntries())
                 {
-                    newTable.RemoveEntries();
+                    if (newTable.TrackIncrementalSteps)
+                    {
+                        newTable.RecordStepInfoForLastEntry(_name, TimeSpan.Zero, ImmutableArray.Create((entry.Step!, entry.OutputIndex)), EntryState.Cached);
+                    }
                 }
-                else if (entry.state != EntryState.Cached || !newTable.TryUseCachedEntries())
+                else
                 {
+                    // start twice to improve accuracy. See AnalyzerExecutor.ExecuteAndCatchIfThrows for more details
+                    _ = SharedStopwatch.StartNew();
+                    var stopwatch = SharedStopwatch.StartNew();
                     // generate the new entries
-                    var newOutputs = _func(entry.item, cancellationToken);
+                    var newOutputs = _func(entry.Item, cancellationToken);
 
-                    if (entry.state != EntryState.Modified || !newTable.TryModifyEntries(newOutputs, _comparer))
+                    if (entry.State != EntryState.Modified || !newTable.TryModifyEntries(newOutputs, _comparer))
                     {
                         newTable.AddEntries(newOutputs, EntryState.Added);
                     }
+
+                    if (newTable.TrackIncrementalSteps)
+                    {
+                        newTable.RecordStepInfoForLastEntry(_name, stopwatch.Elapsed, ImmutableArray.Create((entry.Step!, entry.OutputIndex)), entry.State);
```

Comment:
We could put the name in the table too, as it should be the same for every entry in a given table. Consider doing that with the `CreateTableBuilder` step above?